Update mod to Minecraft 1.21.9

sakura-ryoko opened the pull request:
I also sent a PR for PAPI for PatBox.

sakura-ryoko committed:
Squash.

1.21.9-pre1

Slight cleanup

Bump PAPI, fix Icon Alignment by 2 px

Bump PRE2

Bump PRE3

Bump PRE4

Bump RC.1, address comments.

Bump PAPI

Fix formatting

Update src/main/java/com/terraformersmc/modmenu/gui/widget/entries/ChildEntry.java

Update src/main/java/com/terraformersmc/modmenu/gui/widget/entries/ParentEntry.java

Update src/main/resources/fabric.mod.json

Update src/main/java/com/terraformersmc/modmenu/gui/widget/entries/ParentEntry.java

Update src/main/java/com/terraformersmc/modmenu/gui/widget/ModListWidget.java

Co-Authored-By: haykam821 <24855774+haykam821@users.noreply.github.com>

NotRyken commented on src/main/java/com/terraformersmc/modmenu/gui/widget/ModListWidget.java:
```diff
@@ -256,57 +271,33 @@ protected void drawSelectionHighlight(DrawContext context, int x, int y, int wid
-		int int_5 = MathHelper.floor(y - (double) this.getY()) - this.headerHeight + (int) this.getScrollY() - 4;
-		int index = int_5 / this.itemHeight;
-		return x < (double) this.getScrollbarX() && x >= (double) getRowLeft() && x <= (double) (getRowLeft() + getRowWidth()) && index >= 0 && int_5 >= 0 && index < this.getEntryCount() ? this.children().get(index) : null;
+		return super.getEntryAtPosition(x, y);
```
This causes an infinite loop and SOE when the function is called by [`ParentEntry#isMouseOver`](https://github.com/sakura-ryoko/ModMenu/blob/545dc1ec669f5fdb65bb75067ddd8017101e2e10/src/main/java/com/terraformersmc/modmenu/gui/widget/entries/ParentEntry.java#L201).

It may be possible to simply remove the latter override, I'm not sure.

sakura-ryoko replied:
See if the changes satisfies the request.

NotRyken commented on src/main/java/com/terraformersmc/modmenu/gui/widget/entries/ChildEntry.java:
```diff
@@ -18,26 +19,25 @@ public ChildEntry(Mod mod, ParentEntry parent, ModListWidget list, boolean botto
 	) {
-		super.render(drawContext, index, y, x, rowWidth, rowHeight, mouseX, mouseY, isSelected, delta);
-		x += 4;
+		int x = this.getX() + 4;
```
This should be `parent.getX()`

As it is currently the children rapidly slide offscreen to the right when you expand the parent, which looks hilarious but isn't all that helpful.

sakura-ryoko replied:
See if the changes satisfies the request.

NotRyken reviewed (changes requested):
Found a couple of issues using this while testing my own mods.

NotRyken reviewed (approved):
Appears to work correctly.

haykam821 commented on src/main/java/com/terraformersmc/modmenu/gui/widget/entries/ChildEntry.java:
```diff
@@ -18,26 +19,25 @@ public ChildEntry(Mod mod, ParentEntry parent, ModListWidget list, boolean botto
-	public boolean keyPressed(int keyCode, int scanCode, int modifiers) {
-		if (keyCode == GLFW.GLFW_KEY_LEFT) {
+	public boolean keyPressed(KeyInput input) {
+		if (input.key() == GLFW.GLFW_KEY_LEFT) {
```
```suggestion
		if (input.isLeft()) {
```

haykam821 commented on src/main/java/com/terraformersmc/modmenu/gui/widget/entries/ParentEntry.java:
```diff
@@ -162,24 +161,24 @@ public boolean keyPressed(int keyCode, int scanCode, int modifiers) {
 			list.filter(list.getParent().getSearchInput(), false);
 			return true;
-		} else if (keyCode == GLFW.GLFW_KEY_LEFT) {
+		} else if (input.key() == GLFW.GLFW_KEY_LEFT) {
```
```suggestion
		} else if (input.isLeft()) {
```

haykam821 commented on src/main/java/com/terraformersmc/modmenu/gui/widget/entries/ParentEntry.java:
```diff
@@ -162,24 +161,24 @@ public boolean keyPressed(int keyCode, int scanCode, int modifiers) {
 
 			return true;
-		} else if (keyCode == GLFW.GLFW_KEY_RIGHT) {
+		} else if (input.key() == GLFW.GLFW_KEY_RIGHT) {
```
```suggestion
		} else if (input.isRight()) {
```

haykam821 commented on src/main/resources/fabric.mod.json:
```diff
@@ -24,8 +24,8 @@
-    "fabricloader": ">=0.16.10",
-    "minecraft": "~1.21.6"
+    "fabricloader": ">=0.17.2",
+    "minecraft": ">=1.21.9-rc.1 <=1.21.9"
```
I am not sure of the exact minimum version this pull request would support, but I personally believe ranges should be unbounded for maximum versions:

```suggestion
    "minecraft": ">=1.21.9-"
```

haykam821 commented on src/main/java/com/terraformersmc/modmenu/gui/widget/entries/ParentEntry.java:
```diff
@@ -151,9 +150,9 @@ private void toggleChildren() {
+	public boolean keyPressed(KeyInput input) {
 		String modId = getMod().getId();
-		if (keyCode == GLFW.GLFW_KEY_ENTER || keyCode == GLFW.GLFW_KEY_SPACE) {
+		if (input.key() == GLFW.GLFW_KEY_ENTER || input.key() == GLFW.GLFW_KEY_SPACE) {
```
```suggestion
		if (input.isEnterOrSpace()) {
```

This change is technically a behavioral change by allowing keypad enter to be used. Did that behavior change for vanilla user interface elements in Minecraft 1.21.9?

haykam821 commented on src/main/java/com/terraformersmc/modmenu/gui/widget/ModListWidget.java:
```diff
@@ -248,63 +265,41 @@ protected void renderList(DrawContext drawContext, int mouseX, int mouseY, float
-		if (keyCode == GLFW.GLFW_KEY_UP || keyCode == GLFW.GLFW_KEY_DOWN) {
-			return super.keyPressed(keyCode, scanCode, modifiers);
+	public boolean keyPressed(KeyInput input) {
+		if (input.key() == GLFW.GLFW_KEY_UP || input.key() == GLFW.GLFW_KEY_DOWN) {
```
```suggestion
		if (input.isUp() || input.isDown()) {
```

haykam821 reviewed (changes requested):
The 'Toggle Filter Options' and 'Configure...' buttons do not change the cursor shape unlike other buttons. Links, unlike those generated by commands such as `/tellraw @s {click_event: {action: open_url, url: "https://example.org"}, text: Example}`, also do not change the cursor shape.

In addition, when clicking the 'Website' or 'Issues' buttons and returning to the mods screen, the description of the currently selected mod is not restored.

sakura-ryoko commented:
> The 'Toggle Filter Options' and 'Configure...' buttons do not change the cursor shape unlike other buttons. In addition, when clicking the 'Website' or 'Issues' buttons and returning to the mods screen, the description of the currently selected mod is not restored.

If you can find a reason for the behavior change by all means.

sakura-ryoko committed:
fix: Description list de-sync & add "hover" cursors; URL null handling.

sakura-ryoko committed:
Bump to Release 1.21.9

haykam821 reviewed (changes requested):
The selection highlight for parent mods is misaligned:

<img width="696" height="139" alt="Misaligned selection highlight" src="https://github.com/user-attachments/assets/eda9fc6b-7217-46bf-90ea-c6c9da111ffb" />

In addition, expanding the children of a parent mod incorrectly scrolls the selected mod into view, and child mods are moved more to the right than before.

sakura-ryoko committed:
update: Icon / Text spacing across parent / child mod entries; and on the right side as well.

haykam821 reviewed (changes requested):
The layout remains misaligned compared to the previous Minecraft 1.21.8 version. I recommend running both versions in windowed mode to allow switching between the two layouts to ensure that they match up.

sakura-ryoko committed:
revert text alignment in favor of using getContentY() for desc list;
introduce ModListEntry `offsetY` primarily to "fix" the alignment of the top-most Entry to align with 1.21.8's layout; and then each entry render's with the offset set.  Set to 0 for no Y shift.

sakura-ryoko committed:
notate fix: move "update" icon to match the "Realms" icon's position. (Issue # 769)

sakura-ryoko commented:
> (I also couldn't open an issue on your fork) unsure of who's fault this is, but if this is of any help: I crashed while clicking inside Dynamic FPS' tab I was using this: https://github.com/sakura-ryoko/ModMenu/actions/runs/18168097762 logs: https://mclo.gs/wR5kEYh

That is a problem with the mod that you are clicking on; not Mod Menu's fault.  I have tested this feature with MaLiLib.

adr278 commented:
> > (I also couldn't open an issue on your fork) unsure of who's fault this is, but if this is of any help: I crashed while clicking inside Dynamic FPS' tab I was using this: https://github.com/sakura-ryoko/ModMenu/actions/runs/18168097762 logs: https://mclo.gs/wR5kEYh
> 
> That is a problem with the mod that you are clicking on; not Mod Menu's fault. I have tested this feature with MaLiLib.

oh alright 👍 
thanks for letting me know, much appreciated for the update too ❤️

danielwerg commented:
Thank you updating this!

I had small issue with `+` sign and whatever element it belong too, not sure if this is mod menu issue (nothing happens when you click `+`)

https://github.com/user-attachments/assets/ef240e5f-b875-4be7-852b-85ab535bbcf1

^ [Chat Heads mod](https://modrinth.com/mod/chat-heads) but also same issue applies to [ClickThrough Plus](https://modrinth.com/mod/clickthrough+) as well

sakura-ryoko commented:
> Thank you updating this!
> 
> I had small issue with `+` sign and whatever element it belong too, not sure if this is mod menu issue (nothing happens when you click `+`)
> 2025-10-02_17-46-31.mp4
> 
> ^ [Chat Heads mod](https://modrinth.com/mod/chat-heads) but also same issue applies to [ClickThrough Plus](https://modrinth.com/mod/clickthrough+) as well

The Pull Requests are not for Mod Support; See [Issues](https://github.com/TerraformersMC/ModMenu/issues)

Prospector commented:
awesome!
